flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Backtracking.java

```diff
@@ -0,0 +1,254 @@
+					if (visitedPartition) {
+						availableInputs++;
+					}
+					iter.remove();
+				}
+			}
+			return false;
+		}
+
+		public IntermediateResultPartition getNextRequirement() {
+			return pendingInputs.pop();
+		}
+
+		public boolean needsToBeScheduled() {
+			return numInputs == availableInputs;
+		}
+
+		public void inputFound() {
+			availableInputs++;
+		}
+
+		@Override
+		public String toString() {
+			return "TaskRequirement{" +
+					"executionVertex=" + executionVertex +
+					", pendingInputs=" + pendingInputs.size() +
+					'}';
+		}
+	}
+
+	/**
+	 * Action to be performed on an ExecutionVertex when it is determined to be scheduled.
+	 * @param scheduleAction A ScheduleAction which receives an ExecutionVertex.
+	 */
+	public void setScheduleAction(ScheduleAction scheduleAction) {
+		Preconditions.checkNotNull(scheduleAction);
+		this.scheduleAction = scheduleAction;
+	}
+
+	/**
+	 * Hook executed after backtracking finishes. Note that because of the use of futures, this may
+	 * not be when the scheduleUsingBacktracking() method returns.
+	 * @param postBacktrackingHook A Runnable that is executed after backtracking finishes.
+	 */
+	public void setPostBacktrackingHook(Runnable postBacktrackingHook) {
+		Preconditions.checkNotNull(postBacktrackingHook);
+		this.postBacktrackingHook = postBacktrackingHook;
+	}
+
+	/* Visit the ExecutionGraph from the previously determined sinks using a pre-order depth-first
+	 * iterative traversal.
+	 */
+	public void scheduleUsingBacktracking() {
+
+		while (!taskRequirements.isEmpty()) {
+
+			final TaskRequirement taskRequirement = taskRequirements.peek();
+			final ExecutionVertex task = taskRequirement.getExecutionVertex();
+			task.getCurrentExecutionAttempt().setScheduled();
+
+			if (task.getExecutionState() != ExecutionState.CREATED && task.getExecutionState() != ExecutionState.DEPLOYING) {
+				LOG.debug("Resetting ExecutionVertex {} from {} to CREATED.", task, task.getExecutionState());
+				task.resetForNewExecution();
+				task.getCurrentExecutionAttempt().setScheduled();
+			}
+
+			if (taskRequirement.pendingRequirements()) {
+
+				final IntermediateResultPartition resultRequired = taskRequirement.getNextRequirement();
+
+				if (resultRequired.isLocationAvailable()) {
+					ActorRef taskManager = resultRequired.getLocation().getTaskManager();
+
+					LOG.debug("Requesting availability of IntermediateResultPartition " + resultRequired.getPartitionId());
+					// pin ResulPartition for this intermediate result
+					Future<Object> future = Patterns.ask(
+							taskManager,
+							new TaskMessages.LockResultPartition(
+									resultRequired.getPartitionId(),
+									// We lock this result for all consumers. We have to make sure
+									// to release it once a job finishes.
+									resultRequired.getNumConsumers()
+							),
+							5000 // 5 seconds timeout
+					);
+
+					/** BEGIN Asynchronous callback **/
+					future.onComplete(new OnComplete<Object>() {
```

Comment:
The callback is asynchronous, to prevent the job manager from blocking or waiting in the dispatcher threads. You are right that the locking would be faster to execute in parallel. To not having to deal with the concurrency related issues, I chose not to implement it in parallel. If I think about it now, it is probably not too hard to change that.